warehouse/malware/checks/setup_patterns/check.py

```diff
@@ -0,0 +1,120 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import tarfile
+import zipfile
+
+from io import BytesIO
+from textwrap import dedent
+
+import requests
+import yara
+
+from warehouse.malware.checks.base import MalwareCheckBase
+from warehouse.malware.models import VerdictClassification, VerdictConfidence
+from warehouse.packaging.models import File
+
+
+class SetupPatternCheck(MalwareCheckBase):
+    _yara_rule_file = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), "setup_py_rules.yara"
+    )
+
+    version = 1
+    short_description = "A check for common malicious patterns in setup.py"
+    long_description = dedent(
+        """
+        This check uses YARA to search for common malicious patterns in the setup.py
+        files of uploaded release archives.
+        """
+    )
+    check_type = "event_hook"
+    hooked_object = "File"
+
+    def __init__(self, db):
+        super().__init__(db)
+        self._yara_rules = self._load_yara_rules()
+
+    def _load_yara_rules(self):
+        return yara.compile(filepath=self._yara_rule_file)
+
+    def _extract_setup_contents(self, archive_stream):
```

Comment:
For review: we could probably expose a `malware.checks.services` or similar module somewhere that contains routines like this one, since they'll probably get reused across multiple checks. Thoughts?

Reply:
I'm definitely +1 for setting up some helper methods for common operations like fetching the file from CDN. I'm not sure if we'll have many checks that fetch a specific file aside from setup.py... but a `_extract_file_contents(self, archive_stream, file_path)` type method for doing so would also be pretty rad.

Reply:
> I'm definitely +1 for setting up some helper methods for common operations like fetching the file from CDN. I'm not sure if we'll have many checks that fetch a specific file aside from setup.py... but a `_extract_file_contents(self, archive_stream, file_path)` type method for doing so would also be pretty rad.

Sounds good! I can make a helper for the fetch operation. One challenge with `_extract_file_contents` as a generic helper is that `file_path`'s base directory is indeterminate -- for the `sdist`s that I've observed it tends to be `$packagename-$version/` but I don't believe that's strictly enforced anywhere. We could make "ignores the base directory" an explicit part of the behavior, though.

Reply:
that's absolutely correct, AFAIK there's no enforcement on the name of the base directory so it should probably be presumed guilty of breaking that convention.